File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/ProjectionPruningRule.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.abstractions.FieldId;
import java.util.*;

/** Removes redundant duplicate projection fields while preserving first occurrence order. */
public final class ProjectionPruningRule implements IPlanRewriteRule {
    @Override public String name() { return "projection.pruning"; }
    @Override public List<String> preconditions() { return List.of("plan contains a projection", "projection contains redundant fields"); }
    @Override public List<String> securityObligations() { return List.of("visibility.field", "visibility.relationship", "authorization.required"); }
    @Override public double costImpact() { return .25d; }
    @Override public double benefitEstimate() { return 1.5d; }
    @Override public int priority() { return 20; }
    @Override public boolean canApply(SemanticPlan plan) { Objects.requireNonNull(plan); return containsRedundantProjection(plan.root()); }
    @Override public SemanticPlan apply(SemanticPlan plan) {
        Objects.requireNonNull(plan);
        if (!canApply(plan)) return plan;
        boolean[] changed = {false};
        var root = rewriteNode(plan.root(), changed);
        return changed[0] ? new SemanticPlan(root, plan.requiredSecurityInvariants(), plan.authorizationBinding()) : plan;
    }
    private static SemanticPlanNode rewriteNode(SemanticPlanNode node, boolean[] changed) {
        List<FieldId> fields = dedupe(node.fields(), changed);
        List<SemanticPlanNode> children = new ArrayList<>(node.children().size());
        boolean childChanged = false;
        for (var child : node.children()) { var r = rewriteNode(child, changed); children.add(r); childChanged |= r != child; }
        if (childChanged) changed[0] = true;
        if (fields != node.fields() || childChanged)
            return new SemanticPlanNode(node.id(), node.operation(), node.entityId(), fields, node.viaRelationship(), node.viaConnection(), children,
                    node.queryOptions(), node.authorization(), node.relationshipCardinality(), node.traversalMode(), node.traversalOrder(), node.aggregateExecutionStrategy());
        return node;
    }
    private static List<FieldId> dedupe(List<FieldId> fields, boolean[] changed) {
        if (fields.size() < 2) return fields;
        Set<FieldId> seen = new HashSet<>(); List<FieldId> result = new ArrayList<>(fields.size());
        for (var field : fields) { if (seen.add(field)) result.add(field); else changed[0] = true; }
        return result.size() == fields.size() ? fields : result;
    }
    private static boolean containsRedundantProjection(SemanticPlanNode node) {
        return hasDuplicates(node.fields()) || node.children().stream().anyMatch(ProjectionPruningRule::containsRedundantProjection);
    }
    private static boolean hasDuplicates(List<FieldId> fields) { Set<FieldId> seen = new HashSet<>(); for (var f : fields) if (!seen.add(f)) return true; return false; }
}

File: src/csharp/Foundgine.Core/Semantic/Planning/ProjectionPruningRule.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Removes redundant duplicate projection fields while preserving every field
/// required by the result contract and semantic query evaluation.
///
/// The current SemanticPlan represents requested output and execution fields
/// in one collection. Consequently this rule intentionally does not remove a
/// unique requested field merely because it is not referenced by a predicate.
/// Full dead-field pruning requires a separate requested-vs-working projection
/// representation and is therefore outside this milestone.
/// </summary>
public sealed class ProjectionPruningRule : IPlanRewriteRule
{
    public string Name => "projection.pruning";

    public IReadOnlyList<string> Preconditions =>
        ["plan contains a projection", "projection contains redundant fields"];

    public IReadOnlyList<string> SecurityObligations =>
        ["visibility.field", "visibility.relationship", "authorization.required"];

    public double CostImpact => 0.25d;

    public double BenefitEstimate => 1.5d;

    public bool IsIdempotent => true;

    public int Priority => 20;

    public bool CanApply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        return ContainsRedundantProjection(plan.Root);
    }

    public SemanticPlan Apply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (!CanApply(plan))
            return plan;

        var changed = false;
        var root = RewriteNode(plan.Root, ref changed);
        return changed ? new SemanticPlan(root, plan.RequiredSecurityInvariants, plan.AuthorizationBinding) : plan;
    }

    private static SemanticPlanNode RewriteNode(SemanticPlanNode node, ref bool changed)
    {
        var required = ProjectionPruningRequirements.RequiredRootFields(node);
        var fields = DeduplicatePreservingOrder(node.Fields, required, out var nodeChanged);
        if (nodeChanged)
            changed = true;

        var children = new SemanticPlanNode[node.Children.Count];
        for (var i = 0; i < node.Children.Count; i++)
            children[i] = RewriteNode(node.Children[i], ref changed);

        if (ChildrenChanged(node.Children, children))
            changed = true;

        if (changed && (nodeChanged || !children.SequenceEqual(node.Children)))
            return node with { Fields = fields, Children = children };

        return node;
    }

    private static IReadOnlyList<FieldId> DeduplicatePreservingOrder(
        IReadOnlyList<FieldId> fields,
        IReadOnlySet<FieldId> required,
        out bool changed)
    {
        changed = false;
        if (fields.Count < 2)
            return fields;

        var seen = new HashSet<FieldId>();
        var result = new List<FieldId>(fields.Count);
        foreach (var field in fields)
        {
            // Required is deliberately consulted before removing anything. A
            // required field can only lose redundant occurrences, never its
            // first occurrence.
            if (required.Contains(field) && seen.Add(field))
            {
                result.Add(field);
                continue;
            }

            if (seen.Add(field))
                result.Add(field);
            else
                changed = true;
        }

        return changed ? result : fields;
    }

    private static bool ContainsRedundantProjection(SemanticPlanNode node) =>
        HasDuplicateFields(node.Fields) || node.Children.Any(ContainsRedundantProjection);

    private static bool ChildrenChanged(IReadOnlyList<SemanticPlanNode> before, IReadOnlyList<SemanticPlanNode> after)
    {
        if (before.Count != after.Count)
            return true;

        for (var i = 0; i < before.Count; i++)
            if (!ReferenceEquals(before[i], after[i]))
                return true;

        return false;
    }

    private static bool HasDuplicateFields(IReadOnlyList<FieldId> fields)
    {
        if (fields.Count < 2)
            return false;

        var seen = new HashSet<FieldId>();
        foreach (var field in fields)
            if (!seen.Add(field))
                return true;

        return false;
    }
}